sdk/storage/azure-storage-common/src/main/java/com/azure/storage/common/implementation/FluxInputStream.java

```diff
@@ -0,0 +1,243 @@
+           was emitted by the Flux. */
+        if (this.buffer == null) { // Only executed on first subscription.
+            if (this.lastError != null) {
+                throw logger.logThrowableAsError(this.lastError);
+            }
+            if (this.fluxComplete) {
+                return -1;
+            }
+            throw logger.logExceptionAsError(new IllegalStateException("An unexpected error occurred. No data was "
+                + "read from the stream but the stream did not indicate completion."));
+        }
+
+        /* Now we are guaranteed that buffer is SOMETHING. */
+        /* No data is available in the buffer.  */
+        if (this.buffer.available() == 0) {
+            /* If the flux completed, there is no more data available to be read from the stream. Return -1. */
+            if (this.fluxComplete) {
+                return -1;
+            }
+            /* Block current thread until data is available. */
+            blockForData();
+        }
+
+        /* Data available in buffer, read the buffer. */
+        if (this.buffer.available() > 0) {
+            return this.buffer.read(b, off, len);
+        }
+
+        /* If the flux completed, there is no more data available to be read from the stream. Return -1. */
+        if (this.fluxComplete) {
+            return -1;
+        } else {
+            throw logger.logExceptionAsError(new IllegalStateException("An unexpected error occurred. No data was "
+                + "read from the stream but the stream did not indicate completion."));
+        }
+    }
+
+    @Override
+    public void close() throws IOException {
+        subscription.cancel();
+        if (this.buffer != null) {
+            this.buffer.close();
+        }
+        super.close();
+        if (this.lastError != null) {
+            throw logger.logThrowableAsError(this.lastError);
+        }
+    }
+
+    /**
+     * Request more data and wait on data to become available.
+     */
+    private void blockForData() {
+        lock.lock();
+        try {
+            waitingForData = true;
+            if (!subscribed) {
+                subscribeToData();
+            } else {
+                subscription.request(1);
+            }
+            // Block current thread until data is available.
+            while (waitingForData) {
+                if (fluxComplete) {
+                    break;
+                } else {
+                    try {
+                        dataAvailable.await();
+                    } catch (InterruptedException e) {
+                        throw logger.logExceptionAsError(new RuntimeException(e));
+                    }
+                }
+            }
+        } finally {
+            lock.unlock();
+        }
+    }
+
+    /**
+     * Subscribes to the data with a special subscriber.
+     */
+    private void subscribeToData() {
+        this.data
+            .onBackpressureBuffer(Constants.GB)
```

Comment:
What does it do? 1GB is quite high threshold .

Reply:
This is in case downstream is slower for some reason than the actual upstream flow of data. We dont want to drop any bytebuffers so I think this is the way we tell Reactor don't drop anything. Perhaps the default onBackpressureBuffer() may be more appropriate here than setting it to a GB